src/frontend/src/planner/select.rs

```diff
@@ -166,6 +149,65 @@ impl Planner {
         }
     }
 
+    /// Handle (NOT) EXISTS and (NOT) IN in WHERE clause.
+    ///
+    /// We will use a = b to replace a in (select b from ....) for (NOT) IN thus avoiding adding a
+    /// `LogicalFilter` on `LogicalApply`.
+    fn handle_exists_and_in(
+        &mut self,
+        expr: ExprImpl,
+        negated: bool,
+        input: &mut PlanRef,
+    ) -> Result<()> {
+        let join_type = if negated {
+            JoinType::LeftAnti
+        } else {
+            JoinType::LeftSemi
+        };
+        match expr {
+            ExprImpl::FunctionCall(func_call) if func_call.get_expr_type() == ExprType::In => {
+                let (_, mut inputs, _) = func_call.decompose();
+                assert_eq!(inputs.len(), 2);
+
+                let subquery = inputs.pop().unwrap().into_subquery().unwrap();
+                let is_correlated = subquery.is_correlated();
+
+                let left = inputs.pop().unwrap();
+                let right = InputRef::new(input.schema().fields().len(), subquery.return_type());
+                let eq_cond = FunctionCall::new(ExprType::Equal, vec![left, right.into()])?;
+
+                let right_child = self.plan_query(subquery.query)?.as_subplan();
+
+                *input = Self::create_apply_or_join(
```

Comment:
Variable names are a little bit hard to follow here.
* `inputs`, `left`, `right` are about exprs
* `input`, `right_child` are about plan nodes.

Maybe we can call them `left_expr` (or `operand`), `right_expr`, `right_plan`.

Reply:
or use a block
```rust
let (_, inputs, _) = func_call.decompose();
    let (e, subquery) = inputs.into_iter().collect_tuple().unwrap();
    let subquery = subquery.into_subquery().unwrap();
    let is_correlated = subquery.is_correlated();

    let eq_cond = {
        let left = e;
        let right =
            InputRef::new(input.schema().fields().len(), subquery.return_type());
        FunctionCall::new(ExprType::Equal, vec![left, right.into()])?
    };
```